ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/OptimisticHeadValidator.java

```diff
@@ -107,10 +110,13 @@ private void verifyOptimisticHeads() {
     return executionEngine
         .executePayload(executionPayload)
         .thenAccept(
-            result ->
-                recentChainData
-                    .getForkChoiceStrategy()
-                    .orElseThrow()
-                    .onExecutionPayloadResult(blockRoot, result.getStatus()));
+            result -> {
+              forkChoice.updateLatestValidFinalizedSlot(result);
+
+              recentChainData
+                  .getForkChoiceStrategy()
+                  .orElseThrow()
+                  .onExecutionPayloadResult(blockRoot, result.getStatus());
```

Comment:
hmm, I'm actually thinking it might be best to move all of this into `forkChoice` and do it all on the fork choice thread. `ForkChoiceStrategy` does have locks to make it thread safe but we'd make life a lot easier to reason about if protoarray changes were all made on the `ForkChoice` thread and I think this is the one place we don't respect that currently.